crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2451,6 +2438,39 @@ impl<'db> Class<'db> {
         let scope = self.body_scope(db);
         symbol(db, scope, name)
     }
+
+    /// Return `true` if this class appears to be a cyclic definition,
+    /// i.e., it inherits either directly or indirectly from itself.
+    ///
+    /// A class definition like this will fail at runtime,
+    /// but we must be resilient to it or we could panic.
+    #[salsa::tracked]
```

Comment:
I guess the main cost of this diff (due to the fact that we can't use `specify`) is an additional tracked function for each class. But I think this is a reasonable tradeoff; there are relatively few classes (compared to the number of, say, definitions), so I don't think this will make a huge difference to our total tracked-functions count.

Reply:
Yes. I suppose we could also try not caching this, and just recomputing it everywhere. I don't really have any idea how much the caching helps or hurts us here.

Codspeed's entirely neutral on this PR, so it at least doesn't seem to be leading to a significant regression when checking tomllib. But then tomllib doesn't really have any deep inheritance hierarchies (I don't think!). So it might not be the best benchmark for this specific issue.

Reply:
I think it's fine as is until/unless we have evidence it's an issue.